internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,341 @@
+	}
+	url := *r.proxyURL
+	url.Scheme = "dns"
+	url.Path = "/" + r.proxyURL.Host
+	url.Host = "" // Clear the Host field to conform to the "dns:///" format
+
+	proxyTarget := resolver.Target{URL: url}
+	wcc := &wrappingClientConn{
+		stateListener: r.updateProxyResolverState,
+		parent:        r,
+	}
+	return proxyBuilder.Build(proxyTarget, wcc, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+		r.targetResolver = nil
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+		r.proxyResolver = nil
+	}
+}
+
+// combinedAddressesIfReadyLocked creates a list of combined addresses by
+// pairing each proxy address with every target address. For each pair, it
+// generates a new [resolver.Address] using the proxy address, and adding the
+// target address as the attribute along with user info.
+func (r *delegatingResolver) combinedAddressesIfReadyLocked() ([]resolver.Address, []resolver.Endpoint, bool) {
+	if !r.targetResolverReady || !r.proxyResolverReady {
+		return nil, nil, false
+	}
+
+	// If multiple resolved proxy addresses are present, we send only the
+	// unresolved proxy host and let net.Dial handle the proxy host name
+	// resolution when creating the transport. Sending all resolved addresses
+	// would increase the number of addresses passed to the ClientConn and
+	// subsequently to load balancing (LB) policies like Round Robin, leading
+	// to additional TCP connections. However, if there's only one resolved
+	// proxy address, we send it directly, as it doesn't affect the address
+	// count returned by the target resolver and the address count sent to the
+	// ClientConn.
+	var proxyAddr resolver.Address
+	if len(r.proxyAddrs) == 1 {
+		proxyAddr = r.proxyAddrs[0]
+	} else {
+		proxyAddr = resolver.Address{Addr: r.proxyURL.Host}
+	}
+	var addresses []resolver.Address
+	for _, targetAddr := range r.targetResolverState.Addresses {
+		var user url.Userinfo
+		if r.proxyURL.User != nil {
+			user = *r.proxyURL.User
+		}
+		addresses = append(addresses, proxyattributes.Populate(proxyAddr, proxyattributes.Options{
+			User:        user,
+			ConnectAddr: targetAddr.Addr,
+		}))
+	}
+
+	// Create a list of combined addresses by pairing each proxy endpoint
+	// with every target endpoint. For each pair, it constructs a new
+	// [resolver.Endpoint] using the first address from the proxy endpoint,
+	// as the DNS resolver is expected to return only one address per proxy
+	// endpoint. The target address and user information from the proxy URL
+	// are added as attributes to this address. The resulting list of addresses
+	// is then grouped into endpoints, covering all combinations of proxy and
+	// target endpoints.
+	var endpoints []resolver.Endpoint
```

Comment:
nit: It's good to declare the capacity of the slice here and below.

Reply:
Resolving, we're dealing with infrequent operations here, so avoiding a size hint is up to personal preference.